classes/actions.php

```diff
@@ -207,25 +207,44 @@ public static function duplicate_to_course(array $modules, int $targetcourseid,
 
         $sourcemodinfo = get_fast_modinfo($sourcecourseid);
         $targetmodinfo = get_fast_modinfo($targetcourseid);
+        $targetformat = course_get_format($targetmodinfo->get_course());
+        $targetsectionnum = $targetformat->get_last_section_number();
+
+        // If a new section has been specified, we create one.
+        if ($sectionnum > $targetsectionnum) {
+            if (has_capability('moodle/course:update', context_course::instance($targetcourseid))) {
+                $targetformatopt = $targetformat->get_format_options();
+                // No course format setting or no orphaned sections exist.
+                if (!isset($targetformatopt['numsections']) || !($targetformatopt['numsections'] < $targetsectionnum)) {
+                    course_create_section($targetcourseid);
+                }
 
-        // If a the target section number has been specified and there is no such target section, we do nothing.
-        // This should have been prevented by the UI anyway, but we want to be extra safe here.
-        if ($sectionnum != -1 && $sectionnum > count($targetmodinfo->get_section_info_all()) - 1) {
-            return;
+                // Update course format setting to prevent new orphaned sections.
+                if (isset($targetformatopt['numsections'])) {
+                    update_course((object)array('id' => $targetcourseid, 'numsections' => $targetformatopt['numsections'] + 1));
+                }
+
+                // Make sure new sectionnum is set accurately.
+                $sectionnum = $targetsectionnum + 1;
+            } else {
+                return; // No permissions to add section.
+            }
         }
 
         if ($sectionnum == -1) {
```

Comment:
We're still missing out on the `moodle/course:update` cap check here, right? Or would this work without? I believe it would only work if we had enough sections in the target course?

Reply:
Ah no, we returned before in this case, right? If so, this looks good

Reply:
By this point we should have already confirmed that either they have the permissions to create a section OR no new sections will need to be created.  Once we get to actions.php, if we somehow got here without the correct permissions, I'm not sure what to do except return.  I'll add the check.  They really shouldn't be able to get there, but some people might try to hack through I guess.